packages/slate-react/src/components/android/android-editable.tsx

```diff
@@ -0,0 +1,449 @@
+import React, {
+  useCallback,
+  useEffect,
+  useLayoutEffect,
+  useMemo,
+  useRef,
+  useState,
+} from 'react'
+import {
+  Descendant,
+  Editor,
+  Element,
+  Node,
+  NodeEntry,
+  Range,
+  Text,
+  Transforms,
+} from 'slate'
+import throttle from 'lodash/throttle'
+import scrollIntoView from 'scroll-into-view-if-needed'
+
+import { ReactEditor } from '../..'
+import { ReadOnlyContext } from '../../hooks/use-read-only'
+import { useSlate } from '../../hooks/use-slate'
+import { useIsomorphicLayoutEffect } from '../../hooks/use-isomorphic-layout-effect'
+import {
+  DOMElement,
+  DOMNode,
+  DOMRange,
+  DOMStaticRange,
+  getDefaultView,
+  isDOMNode,
+  isPlainTextOnlyPaste,
+} from '../../utils/dom'
+import {
+  EDITOR_TO_ELEMENT,
+  EDITOR_TO_RESTORE_DOM,
+  EDITOR_TO_WINDOW,
+  ELEMENT_TO_NODE,
+  IS_FOCUSED,
+  IS_READ_ONLY,
+  NODE_TO_ELEMENT,
+  PLACEHOLDER_SYMBOL,
+} from '../../utils/weak-maps'
+import { AndroidInputManager } from './android-input-manager'
+import { EditableProps } from '../editable'
+import { ErrorBoundary } from './ErrorBoundary'
+import useChildren from '../../hooks/use-children'
+import {
+  defaultDecorate,
+  hasEditableTarget,
+  hasTarget,
+  isEventHandled,
+  isTargetInsideVoid,
+} from '../default-editable'
+
+const HAS_BEFORE_INPUT_SUPPORT = true
+
+export const AndroidEditableNoError = (props: EditableProps): JSX.Element => {
+  return (
+    <ErrorBoundary>
+      <AndroidEditable {...props} />
+    </ErrorBoundary>
+  )
+}
+
+/**
+ * Editable.
+ */
+
+export const AndroidEditable = (props: EditableProps): JSX.Element => {
+  const {
+    autoFocus,
+    decorate = defaultDecorate,
+    onDOMBeforeInput: propsOnDOMBeforeInput,
+    placeholder,
+    readOnly = false,
+    renderElement,
+    renderLeaf,
+    style = {},
+    as: Component = 'div',
+    ...attributes
+  } = props
+  const editor = useSlate()
+  const ref = useRef<HTMLDivElement>(null)
+  const inputManager = useMemo(() => new AndroidInputManager(editor), [editor])
+
+  // Update internal state on each render.
+  IS_READ_ONLY.set(editor, readOnly)
+
+  // Keep track of some state for the event handler logic.
+  const state = useMemo(
+    () => ({
+      isComposing: false,
+      isUpdatingSelection: false,
+      latestElement: null as DOMElement | null,
+    }),
+    []
+  )
+
+  // Update element-related weak maps with the DOM element ref.
+  useIsomorphicLayoutEffect(() => {
```

Comment:
Hi @wleroux , there are a number of these `useIsomorphicLayoutEffect` methods that are verbatim from `editable.tsx`.

I wonder if it might be better to take some of these, especially the longer and more complex methods, and turn them into a custom hook that is exported from `editable.tsx`. Maybe this particular one isn't a great example (because it's not that long), but as an example, it could be something like: `useUpdateWeakMaps(editor, window, ref)`.

By having them in a separate `AndroidEditable` still gives us the option to "roll back" just the AndroidEditable version if the default version causes problems specific to Android.

I won't add comments to each one but just as a general idea.

Reply:
`Editable.tsx` is a very frequently changed file and all the effort into making this portion re-usable will have to be re-done every time `Editable.tsx` changes in the main branch.

I would like to deal with this after the pull request is merged. By delaying it until after the merge, the pull request would be a trivial refactoring. Alternatively, we can delay this change until right before this pull request merge. Thoughts?

Reply:
Yep, that's fine to defer IMO as long as we do it.